overlord/snapstate/snapstate.go

```diff
@@ -538,6 +539,11 @@ func InstallPath(st *state.State, si *snap.SideInfo, path, instanceName, channel
 		}
 	}
 
+	channel, err = withPinnedTrack(st, instanceName, channel)
```

Comment:
note that at this point AFAICT the channel has not been through `snap.ParseChannel`, so it's not been through `(snap.Channel).Clean`, so it's not been normalised. So it can be for example `"stable/hotfix-1234"`, which resolves to `"latest/stable/hotfix-1234"` currently.

Because `snap.ParseChannel` does not set `Track` unless the user specified it explicitly, calling it before `withPinnedTrack` and passing it in (to both `withPinnedTrack` and `canSwitchChannel` while you're at it) would actually simplify the code.

Alternatively, parse the channel in `canSwitchChannel`, and call `withPinnedTrack` from there?

Reply:
I've cleaned this up a bit rolled 2 separate funcs into one that computes the effective channel to use. While at it, I've added more tests. Needs another look now.